### PR TITLE
Add Windows CMake CI

### PR DESCRIPTION
**Pull request by duncanspumpkin**
I dont really know what I'm doing but this look to compile on the CI. One thing to note is its not caching anything so wastes lots of time on vcpkg stuff.

### CODE REVIEW

**Comment by IntelOrca** [comment]
I suggest using ninja over msbuild. And we can also use a vcpkg.json file so you don't need to do any explicit install beforehand.

https://github.com/IntelOrca/libsawyer/blob/master/.github/workflows/ci.yml#L71
https://github.com/IntelOrca/libsawyer/blob/master/vcpkg.json

**Comment by AaronVanGeffen** [comment]
Would it be possible to reuse the vcpkg dependency bundle instead of recompiling them?

**Comment by duncanspumpkin** [comment]
Okay so i switched it to using the vcpkg.json and now the macosx build is failing. Also i can't work out where it's actually building the files to

**Comment by IntelOrca** [comment]
> Also i can't work out where it's actually building the files to

Well it is where you specify for the cmake config. `./build` in this case.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -54,6 +54,42 @@ jobs:
...
+        shell: cmd
+        run: |
+            mkdir build
+            call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\Tools\VsDevCmd.bat" -arch=x86
+            cmake -G Ninja -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=%VCPKG_INSTALLATION_ROOT%\scripts\buildsystems\vcpkg.cmake
+            cmake --build build
+      - name: Build artifacts
+        run: |
+            mkdir artifacts | Out-Null
+            Copy-Item CHANGELOG.md,CONTRIBUTORS.md,LICENSE,loco.exe,.build\Release\*.dll,.build\Release\*.pdb artifacts
```
Comment (IntelOrca) [inline .github/workflows/ci.yml:84]:
```suggestion
            Copy-Item CHANGELOG.md,CONTRIBUTORS.md,LICENSE,loco.exe,build\Release\*.dll,build\Release\*.pdb artifacts
```

File: vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "openloco",
+    "version-string": "21.08",
+    "dependencies": [
+        "breakpad",
+        "sdl2",
+	"sdl2-mixer",
+	"zlib",
+	"libpng",
+	"yaml-cpp"
```
Comment (IntelOrca) [inline vcpkg.json:10]: Incorrect indentation

**Comment by duncanspumpkin** [comment]
I give up I just cannot work out how you are meant to cache vcpkg or how to prevent the mac build failing.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -54,6 +54,47 @@ jobs:
...
           if-no-files-found: error
+  windows2:
+    name: Windows Cmake
+    runs-on: windows-latest
+    env:
+      CONFIGURATION: Release
+      POWERSHELL_TELEMETRY_OPTOUT: 1
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (Margen67) [inline .github/workflows/ci.yml:65]:
```suggestion
        uses: actions/checkout@v3
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -54,6 +54,47 @@ jobs:
...
+        with:
+          fetch-depth: 0
+      - name: Prepare cache key
+        id: key
+        shell: powershell
+        run: |
+          New-Item -Type Junction -Path vcpkg -Target c:\vcpkg
```
Comment (Margen67) [inline .github/workflows/ci.yml:72]:
```suggestion
        run: New-Item -Type Junction -Path vcpkg -Target c:\vcpkg
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -54,6 +54,47 @@ jobs:
...
+      - name: vcpkg Cache
+        uses: actions/cache@v2
```
Comment (Margen67) [inline .github/workflows/ci.yml:74]:
```suggestion
        uses: actions/cache@v3
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -54,6 +54,47 @@ jobs:
...
+        with:
+          path: |
+            vcpkg/installed
+          key: x86-vcpkg-${{ hashFiles('vcpkg.json') }}
+      - name: Build OpenLoco
+        shell: cmd
+        run: |
+            mkdir build
```
Comment (Margen67) [inline .github/workflows/ci.yml:82]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -54,6 +54,47 @@ jobs:
...
+            cmake -G Ninja -A Win32 -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=%VCPKG_INSTALLATION_ROOT%\scripts\buildsystems\vcpkg.cmake
+            cmake --build build
+      - name: Build artifacts
+        run: |
+            mkdir artifacts | Out-Null
+            Copy-Item CHANGELOG.md,CONTRIBUTORS.md,LICENSE,loco.exe,build\*.dll,build\*.pdb artifacts
+            Copy-Item data\language artifacts\data\language -Recurse
+            Rename-Item artifacts\loco.exe openloco.exe
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v2
```
Comment (Margen67) [inline .github/workflows/ci.yml:93]:
```suggestion
        uses: actions/upload-artifact@v3
```

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -56,6 +56,41 @@ jobs:
...
+            mkdir artifacts | Out-Null
+            Copy-Item CHANGELOG.md,CONTRIBUTORS.md,LICENSE,loco.exe,build\*.dll,build\*.pdb artifacts
+            Copy-Item data\language artifacts\data\language -Recurse
+            Rename-Item artifacts\loco.exe openloco.exe
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: OpenLoco-${{ runner.os }}-Win32-cmake
+          path: artifacts
+          if-no-files-found: error
```
Comment (Margen67) [inline .github/workflows/ci.yml:93]:
```suggestion
          if-no-files-found: error
```

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -56,6 +56,41 @@ jobs:
...
+        run: |
+          call "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\VsDevCmd.bat" -arch=x86
+          cmake -G "Ninja" -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=${{ github.workspace }}\vcpkg\scripts\buildsystems\vcpkg.cmake
+          cmake --build build
+      - name: Build artifacts
+        run: |
+            mkdir artifacts | Out-Null
+            Copy-Item CHANGELOG.md,CONTRIBUTORS.md,LICENSE,loco.exe,build\*.dll,build\*.pdb artifacts
+            Copy-Item data\language artifacts\data\language -Recurse
+            Rename-Item artifacts\loco.exe openloco.exe
```
Comment (Margen67) [inline .github/workflows/ci.yml:87]:
```suggestion
          mkdir artifacts | Out-Null
          Copy-Item CHANGELOG.md,CONTRIBUTORS.md,LICENSE,loco.exe,build\*.dll,build\*.pdb artifacts
          Copy-Item data\language artifacts\data\language -Recurse
          Rename-Item artifacts\loco.exe openloco.exe
```

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -56,6 +56,41 @@ jobs:
...
+      CONFIGURATION: Release
+      POWERSHELL_TELEMETRY_OPTOUT: 1
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: true
+      - name: Restore artifacts, or setup vcpkg
+        uses: lukka/run-vcpkg@v10
```
Comment (Margen67) [inline .github/workflows/ci.yml:72]:
```suggestion
        uses: lukka/run-vcpkg@v10.4
```

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -56,6 +56,41 @@ jobs:
...
+        with:
+          vcpkgDirectory: '${{ github.workspace }}/vcpkg'
+          vcpkgGitCommitId: 01b29f6d8212bc845da64773b18665d682f5ab66 
+      - name: Build OpenLoco
+        shell: cmd
+        run: |
+          call "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\VsDevCmd.bat" -arch=x86
+          cmake -G "Ninja" -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=${{ github.workspace }}\vcpkg\scripts\buildsystems\vcpkg.cmake
```
Comment (Margen67) [inline .github/workflows/ci.yml:80]:
```suggestion
          call "%ProgramFiles%\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\VsDevCmd.bat" -arch=x86
          cmake -GNinja -Bbuild -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=${{ github.workspace }}\vcpkg\scripts\buildsystems\vcpkg.cmake
```

**Review by AaronVanGeffen — APPROVED** [review]
Looks good to me, but haven't been able to test a Win32 CMake artefact. Would appreciate a review from someone else who can do this as well.

File: vcpkg.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "openloco",
+    "version-string": "22.02",
```
[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply vcpkg.json:3]: It's just a version string for our dependencies. We last updated them with 22.02

**Comment by duncanspumpkin** [comment]
> Looks good to me, but haven't been able to test a Win32 CMake artefact. Would appreciate a review from someone else who can do this as well.

Tested the artifact and noticed i wasn't actually outputing anything after some directory changes. Woops. All fixed now. Also got it building in almost the exact same time.

**Pull request merged**